### PR TITLE
Tokens serialization

### PR DESCRIPTION
**Pull request by dks17**
This is custom [coder](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html#method-i-serialize) for tokens column with known object's structure (only Hash). This is based on `JSON` coder with some edge cases (nil, blank string).

`PG` registers `jsonb` as alias type for `json` and uses own [text coder](https://github.com/ged/ruby-pg/blob/b920bb33eaa4797cdfdca8885dd56dbd04d6b8ea/lib/pg/basic_type_mapping.rb#L218-L219) where also uses `JSON` coder and few other date/time formats for [encoding](https://github.com/ged/ruby-pg/blob/1d9883d06bdf5413880f20bbb445720dedebd736/lib/pg/text_encoder.rb)/[decoding](https://github.com/ged/ruby-pg/blob/1d9883d06bdf5413880f20bbb445720dedebd736/lib/pg/text_decoder.rb).

Should help to serialize/deserialize tokens properly.

This should fix:
#1079 #121
and may:
#1240 #1224 

Need help testing the PR before it will be merged.

**UPDATE**

- Updated `dump` method.
- Deleted code from `load` method that recognize case when json is blank string (it is unlikely case).

### CODE REVIEW

File: app/models/devise_token_auth/concerns/tokens_serialization.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module DeviseTokenAuth::TokensSerialization
+  # Serialization hash to json
+  def self.dump(object)
+    object.nil? ? '{}' : object.each_value(&:compact!).to_json
```
Comment (dks17) [inline app/models/devise_token_auth/concerns/tokens_serialization.rb]: I noticed that value of `last_token` very often equals nil (it seems batch) while `updated_at` is still present and updated. So delete (`&:compact!`) all keys with nil value.

Comment (dks17) [inline app/models/devise_token_auth/concerns/tokens_serialization.rb]: We could store 'null' instead of '{}' to `NULL` operators could be used.

**Comment by dks17** [comment]
@MaicolBen Wrote small description about it at the top message.

File: app/models/devise_token_auth/concerns/tokens_serialization.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+module DeviseTokenAuth::TokensSerialization
+  # Serialization hash to json
+  def self.dump(object)
+    object.each_value(&:compact!) unless object.nil?
```
Comment (dks17) [inline app/models/devise_token_auth/concerns/tokens_serialization.rb]: If we care about storing `null` values we can delete this line.
See `last_token` below:

"{\"nlw47IfMSJxHl8v2bvhrOw\":{\"token\":\"$2a$10$3yWVUhPtnn/vigTztah2fuioMMEB6R8aZunkFOjzguV1ZCZ8kga16\",\"expiry\":1549026407,\"last_token\":null,\"updated_at\":\"2019-01-18T13:06:32.876Z\"}}"

File: app/models/devise_token_auth/concerns/tokens_serialization.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    JSON.generate(object)
```
Comment (dks17) [inline app/models/devise_token_auth/concerns/tokens_serialization.rb]: Serialize nil to `null`:
```ruby
JSON.generate(nil)
=> "null"
```

File: app/models/devise_token_auth/concerns/tokens_serialization.rb (hunk 1 above)
Comment (dks17) [reply app/models/devise_token_auth/concerns/tokens_serialization.rb]: Anyway `nil` or `{}` is stored as `null`.

**Comment by kdubel** [comment]
Thanks for this fix! I had v1.1.0 of gem and the problem existed, updated to your fix and everything works fine. @MaicolBen maybe it's a good idea to bump version as searching for this issue takes some time?

**Comment by softwaregravy** [comment]
I'm tracing through the threads on this issue as well. I agree it would be great to get this out.

**Pull request merged**